boxoffice/models/invoice.py

```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
+
+from datetime import datetime
+from decimal import Decimal
+from boxoffice.models import db, BaseMixin, UuidMixin
+from sqlalchemy import event
+from sqlalchemy import sql, func
+from sqlalchemy.ext.orderinglist import ordering_list
+from boxoffice.models.user import get_financial_year
+
+__all__ = ['Invoice']
+
+
+def get_latest_invoice_no(organization, jurisdiction, invoice_dt):
+    """
+    Returns the last invoice number used, 0 if no order has been invoiced yet
+    """
+    query = db.session.query(sql.functions.max(Invoice.invoice_no))\
+        .filter(Invoice.organization == organization)
+    fy_start_at, fy_end_at = get_financial_year(jurisdiction, invoice_dt)
+    if fy_start_at and fy_end_at:
+        query = query.filter(Invoice.invoiced_at >= fy_start_at, Invoice.invoiced_at < fy_end_at)
+    return query.scalar() or 0
+
+
+class Invoice(UuidMixin, BaseMixin, db.Model):
+    __tablename__ = 'invoice'
+    __uuid_primary_key__ = True
+    __table_args__ = (db.UniqueConstraint('organization_id', 'invoice_no'),)
+
+    invoicee_name = db.Column(db.Unicode(255), nullable=True)
+    invoicee_email = db.Column(db.Unicode(254), nullable=True)
+    invoice_no = db.Column(db.Integer(), nullable=True)
+    invoiced_at = db.Column(db.DateTime, nullable=True)
+    street_address_1 = db.Column(db.Unicode(255), nullable=True)
+    street_address_2 = db.Column(db.Unicode(255), nullable=True)
+    city = db.Column(db.Unicode(255), nullable=True)
+    state = db.Column(db.Unicode(255), nullable=True)
+    # India specific: this is the state short code. Eg: KA for Karnataka
+    state_code = db.Column(db.Unicode(4), nullable=True)
```

Comment:
This is not India specific. We _currently_ only need it for India. Use [ISO 3166-2 codes](https://en.wikipedia.org/wiki/ISO_3166-2), which require _up to_ three digits (not 4).